### ISSUE TITLE
Softmax stamp for image in detail view

### ISSUE THREAD

**Issue by thinkh** [Caleydo/malevo#27]
- Selected image
- Epochs over time (x-axis)
- Classes (y-axis)
- Softmax value (color brightness)

**Comment by thinkh** [comment]
@NMO13 please add a sketch

**Comment by mstreit** [comment]
![ink-image](https://user-images.githubusercontent.com/4129755/36417327-ed666d72-162b-11e8-996f-d97f1d6afc01.png)

**Comment by mstreit** [comment]
also see sketch on slide 46 in https://docs.google.com/presentation/d/1G5E-HJ_PRDsJ90YH6vd2Kf4L6wR-_SjkRns3BUD8u5g/edit#slide=id.g3314dfedfb_0_0

![image](https://user-images.githubusercontent.com/4129755/36430448-7f56a8f4-1655-11e8-8b45-944fa71bfa1d.png)

**Comment by thinkh** [comment]
@NMO13 You can start with dummy data until the API returns real softmax data. The API is defined in the [swagger.yml](https://github.com/Caleydo/malevo/blob/gfrogat/features/api/malevo/swagger.yml#L186).

**Comment by thinkh** [comment]
@NMO13 If possible, use the [Phovea heatmap](https://github.com/phovea/phovea_vis/tree/develop/src/heatmap). You can see the [usage of the heatmap in TACO](https://github.com/Caleydo/taco/blob/master/src/heat_map.ts).

**Comment by gfrogat** [comment]
I'm aware that we limited the scope of `malevo` to the 10 class problem, but the softmax stamp approach won't scale to settings where we have (much) more than 10 classes. I doubt that we'll be able to see anything in these large stamps --> maybe we should already think about other marks.

aside from the perceptual issues, we're already running into scalability problems on the backend. I setup a first database with the softmax values in long format for one (1) run and I'm already around 2GB (only considering every 5 epochs). If we have more epochs or more classes the database sizes will explode.

The question is now, do we really need the softmax values or can we get away by only using the loss value for all instances (e.g. cross entropy) and visualize how the loss per instance changes over the training.

The big advantage would be that it's more general as it will work with any type of loss function and we need way less data (only one tenth for cifar10 and even way less for datasets with more classes).

By looking directly at the loss we could even detect instances that mess up the learning during training.

Any thoughts?

**Comment by gfrogat** [comment]
## softmax stamps - Adam
![softmax_stamp_resnet34_adam](https://user-images.githubusercontent.com/11089303/37638285-593c1fe2-2c0c-11e8-8a4d-da3e2d3000e2.png)

## instance loss - Adam
![instance_loss_resnet34_adam](https://user-images.githubusercontent.com/11089303/37638296-64e1746e-2c0c-11e8-9427-445ed5e4d329.png)

## number of misclassifications over epochs - Adam
![no_misclassifications_resnet34_adam](https://user-images.githubusercontent.com/11089303/37638297-667d762e-2c0c-11e8-96e3-fa3ad8806ca7.png)

## softmax stamps and instance loss - SGD
![softmax_stamp_resnet34_sgd](https://user-images.githubusercontent.com/11089303/37638312-758cb580-2c0c-11e8-847d-9c0ef6ce43bb.png)

## instance loss - SGD
![instance_loss_resnet34_sgd](https://user-images.githubusercontent.com/11089303/37638316-78ec34bc-2c0c-11e8-9b9c-e799a255ab80.png)

## number of misclassifications over epochs - SGD
![no_misclassifications_resnet34_sgd](https://user-images.githubusercontent.com/11089303/37638318-7b51bcf4-2c0c-11e8-933a-7aaac6439e18.png)

**Comment by gfrogat** [comment]
the softmax stamp won't work out of the box if we aggregate multiple classes to groups (they also don't make much sense for groups of classes).

It's also not immediately clear how we could aggregate the softmax stamps for groups. Some possibilities I thought of:
- weighted sum / average of all softmax probabilities per group
- compute within group differences (e.g. using cross entropy)


We could also pursue a multi-level approach - first visualize "between-group softmax" and then zoom into "within-group softmax"

But it would be better to come up with a visualization which supports groups and classes natively.

**Comment by NMO13** [comment]
That's true but I think we have the same kind of problems for other detail views as well.
We have to redesign the detail views for hierarchical classes / classes that have more than 10 classes. I will create an issue for that.

**Comment by thinkh** [comment]
To discuss: How to change all detail views to apply the hierarchical structure?

**Comment by thinkh** [comment]
Out of scope for the paper. Hence, moved to icebox.
